File: Other Problems/minKnightMovesBishop.py
from collections import deque

def isOnBishop(x, y, bishopRow, bishopCol):
  return abs(y - bishopCol) == abs(bishopRow - x)

def minKnightMovesBishop(n, rowStart, colStart, rowEnd, colEnd, bishopRow, bishopCol):
  queue = deque([[rowStart, colStart, False]])
  bishopAliveTracker = {}
  bishopAliveTracker[str([rowStart, colStart])] = 1
  count = 0
  bishopDeadTracker = {}

  xDir = [-2, -2, -1, -1, 1, 1, 2, 2]
  yDir = [-1, 1, -2, 2, -2, 2, -1, 1]

  while len(queue) != 0:
    queueLength = len(queue)

    for i in range(queueLength):
      currentPos = queue.popleft()
      if currentPos[0] == rowEnd and currentPos[1] == colEnd:
        return count

      for j in range(len(xDir)):
        newX = currentPos[0] + xDir[j]
        newY = currentPos[1] + yDir[j]
        bishopIsDead = ((newX == bishopRow and newY == bishopCol) or currentPos[2])
        newCoord = [newX, newY, bishopIsDead]

        if not bishopIsDead:
          if str(newCoord) not in bishopAliveTracker and newX < n and newY < n and newX >= 0 and newY >= 0 and not isOnBishop(newX, newY, bishopRow, bishopCol):
            queue.append(newCoord)
            bishopAliveTracker[str(newCoord)] = 1
        else:
          if str(newCoord) not in bishopDeadTracker and newX < n and newY < n and newX >= 0 and newY >= 0:
            queue.append(newCoord)
            bishopDeadTracker[str(newCoord)] = 1
    count += 1
  return -1

print(minKnightMovesBishop(150, 120, 131, 149, 149, 140, 140))

File: Other Problems/minKnightMovesBishop.js
var isOnBishop = function(x, y, bishopRow, bishopCol) {
  return Math.abs(y - bishopCol) === Math.abs(bishopRow - x);
}

var minKnightMovesBishop = function(n, startRow, startCol, endRow, endCol, bishopRow, bishopCol) {
  var tracker = {};
  tracker[[startRow, startCol]] = 1;

  var bishopTracker = {};
  var count = 0;

  var xDir = [-2, -2, -1, -1, 1, 1, 2, 2];
  var yDir = [-1, 1, -2, 2, -2, 2, -1, 1];

  var queue = [[startRow, startCol, false]]; //isBishopDead
  if (bishopRow === startRow && bishopCol === startCol) {
    queue[0][2] = true;
  } else {
    tracker[[bishopRow, bishopCol]] = 1;
  }
  debugger;
  while (queue.length !== 0) {
    var queueLength = queue.length;

    for (var i = 0; i < queueLength; i++) {
      var currentPos = queue.shift();
      if (currentPos[0] === endRow && currentPos[1] === endCol) {
        return count;
      }

      for (var j = 0; j < xDir.length; j++) {
        var newX = currentPos[0] + xDir[j];
        var newY = currentPos[1] + yDir[j];
        var bishopIsDead = ((newX === bishopRow && newY === bishopCol) || currentPos[2]);
        var newCoord = [newX, newY, bishopIsDead];

        if (!bishopIsDead) {
          if (tracker[newCoord] === undefined && newX < n && newY < n && newX >= 0 && newY >= 0 && !isOnBishop(newX, newY, bishopRow, bishopCol)) {
            queue.push(newCoord);
            tracker[newCoord] = 1;
          }
        } else {
          if (bishopTracker[newCoord] === undefined && newX < n && newY < n && newX >= 0 && newY >= 0) {
            queue.push(newCoord);
            bishopTracker[newCoord] = 1;
          }
        }
      }
    }
    count++;
  }
  return -1;
}

console.log(minKnightMovesBishop(150, 120, 130, 149, 149, 140, 140));

